File: src/OTAPackage.java
/*
 * OTA Catalog Parser
 * Copyright (c) 2016 Dialexio
 * 
 * The MIT License (MIT)
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import com.dd.plist.*;
import java.text.NumberFormat;
import java.util.Locale;
import java.util.regex.*;

class OTAPackage {
	private final long SIZE;
	private final NSDictionary ENTRY;
	private final String URL;
	private Matcher match;

	public final static String REGEX_BETA = "(\\d)?\\d[A-Z][4-6]\\d{3}[a-z]?";

	public OTAPackage(NSDictionary otaEntry) {
		ENTRY = otaEntry;
		otaEntry = null;

		// Obtaining the size and URL.
		// First, we need to make sure we don't get info for a dummy update.
		if (ENTRY.containsKey("RealUpdateAttributes")) {
			NSDictionary REAL_UPDATE_ATTRS = (NSDictionary)ENTRY.get("RealUpdateAttributes");

			SIZE = ((NSNumber)REAL_UPDATE_ATTRS.get("RealUpdateDownloadSize")).longValue();
			URL = REAL_UPDATE_ATTRS.get("RealUpdateURL").toString();
			REAL_UPDATE_ATTRS = null;
		}

		else {
			SIZE = ((NSNumber)ENTRY.get("_DownloadSize")).longValue();
			URL = ((NSString)ENTRY.get("__BaseURL")).getContent() + ((NSString)ENTRY.get("__RelativePath")).getContent();
		}
	}

	/**
	 * Returns the actual build number of the OTA entry.
	 * Sometimes, Apple likes to add a large number to the end.
	 * 
	 * @return The build number that iOS will report in Settings.
     **/
	public String actualBuild() {
		// If it the build number looks like a beta...
		// And it's labeled as a beta...
		// But it's not a beta... We need the actual build number.
		if (this.declaredBuild().matches(REGEX_BETA) &&
			this.releaseType().equals("Public") == false &&
			this.actualReleaseType() == 0) {
				int letterPos, numPos;

				for (letterPos = 1; letterPos < this.declaredBuild().length(); letterPos++) {
					if (Character.isUpperCase(this.declaredBuild().charAt(letterPos))) {
						letterPos++;
						break;
					}
				}

			    numPos = letterPos + 1;
				if (this.declaredBuild().charAt(numPos) == '0')
					numPos++;

				return this.declaredBuild().substring(0, letterPos) + this.declaredBuild().substring(numPos);
		}
		
		else
			return this.declaredBuild();
	}

	/**
	 * Returns the package's actual release type in the form of an integer.
	 * 
	 * @return An integer value of 0 (not a beta), 1 (public beta), 2 (developer beta), 3 (carrier beta), or 4 (internal build). -1 may be returned if the type is unknown.
     **/
	public int actualReleaseType() {
		// Just check ReleaseType and return values based on it.
		// We do need to dig deeper if it's "Beta" though.
		if (this.releaseType().equals("Public"))
			return 0;

		else {
			switch (ENTRY.get("ReleaseType").toString()) {
				case "Beta":
					if (this.documentationID().equals("N/A") || this.documentationID().equals("PreRelease"))
						return 2;

					else if (this.documentationID().contains("Public"))
						return 1;

					else if (this.documentationID().contains("Beta") || this.documentationID().contains("Seed"))
						return 2;

					else
						return 0;

				case "Carrier":
					return 3;

				case "Internal":
					return 4;

				default:
					System.err.println("Unknown ReleaseType: " + ENTRY.get("ReleaseType").toString());
					return -1;
			}
		}
	}

	/**
	 * Returns the beta number of this entry.
	 * If a beta number cannot be determined, returns 0.
	 * 
	 * @return Whatever number beta this is, as an int.
     **/
	public int betaNumber() {
		final char digit = this.documentationID().charAt(this.documentationID().length() - 1);


		if (this.isHonestBuild() && (this.documentationID().contains("Public") || this.documentationID().contains("Beta") || this.documentationID().contains("Seed")))
			return (Character.isDigit(digit)) ? Character.getNumericValue(digit) : 1;

		else
			return 0;
	}

	/**
	 * Returns the value in the entry's CompatibilityVersion key.
	 * If the compatibility version cannot be determined, this returns 0.
	 * 
	 * @return Whatever is in the key CompatibilityVersion, as an int.
     **/
	public int compatibilityVersion() {
		return (ENTRY.containsKey("CompatibilityVersion")) ? (int)(ENTRY.get("CompatibilityVersion").toJavaObject()) : 0;
	}

	/**
	 * Returns the timestamp found in the URL.
	 * Note that this is not always accurate;
	 * it may be off by a day, or even a week.
	 * 
	 * @return The timestamp found in the URL, which may not be accurate.
     **/
	public String date() {
		match = Pattern.compile("\\d{4}(\\-|\\.)20\\d{8}\\-").matcher(URL);

		if (match.find())
			return match.group().substring(5, 9) + match.group().substring(10, 12) + match.group().substring(13, 15);

		else {
			match = Pattern.compile("\\d{4}(\\-|\\.)20\\d{4}(\\d|\\.)(\\w|\\-)").matcher(URL);

			if (match.find()) {
				switch (match.group().substring(5)) {
					case "201218.D22":
						return "20120307";

					case "2015106-DC":
						return "20151006";

					case "20160009/1":
						return "20160913";

					default:
						return match.group().substring(5);
				}
			}
	
			else
				return "00000000";
		}
	}

	/**
	 * Returns part of the timestamp found in the URL.
	 * 
	 * @param dmy	Specifies if you are looking for the <b>d</b>ay, <b>m</b>onth, or <b>y</b>ear.
	 * If neither 'd', 'm', or 'y' is specified, it will return the entire timestamp.
	 * 
	 * @return The day, month, or year found in the URL.
     **/
	public String date(final char dmy) {
		switch (dmy) {
			// Day
			case 'd':
				return this.date().substring(6);

			// Month
			case 'm':
				return this.date().substring(4, 6);

			// Year
			case 'y':
				return this.date().substring(0, 4);

			default:
				return this.date();
		}
	}

	/**
	 * @return The build as listed in the OTA update catalog.
     **/
	public String declaredBuild() {
		return ENTRY.get("Build").toString();
	}

	/**
	 * Reports the documentation ID that Apple assigned.
	 * This tells iOS which documentation file to load, since multiple ones may be listed.
	 * 
	 * @return The documentation ID that corresponds to the OTA update. If one is not specified, returns "N/A."
     **/
	public String documentationID() {
		return ENTRY.containsKey("SUDocumentationID") ? ENTRY.get("SUDocumentationID").toString() : "N/A";
	}

	/**
	 * Checks if the release has an inflated build number.
	 * Apple does this to push devices on beta builds to stable builds.
	 * 
	 * @return A boolean value of whether this release has a false build number (true) or not (false).
     **/
	public boolean isHonestBuild() {
		return this.actualBuild().equals(this.declaredBuild());
	}

	/**
	 * Checks if the release is a large, "one size fits all" package.
	 * 
	 * @return A boolean value of whether this release is used to cover all scenarios (true) or not (false).
     **/
	public boolean isUniversal() {
		return (this.prerequisiteBuild().equals("N/A"));
	}

	/**
	 * Returns the value of "MarketingVersion" if present.
	 * "MarketingVersion" is used in some entries to display
	 * a false version number.
	 * 
	 * @return A String value of the "MarketingVersion" key (if it exists), otherwise returns the "OSVersion" key.
     **/
	public String marketingVersion() {
		if (ENTRY.containsKey("MarketingVersion")) {
			if (ENTRY.get("MarketingVersion").toString().contains(".") == false)
				return ENTRY.get("MarketingVersion").toString() + ".0";
			else
				return ENTRY.get("MarketingVersion").toString();
		}

		else
			return this.osVersion();
	}

	/**
	 * @return The "OSVersion" key, as a String.
     **/
	public String osVersion() {
		String version = ENTRY.get("OSVersion").toString();

		return (version.substring(0, 3).equals("9.9")) ? version.substring(4) : version;
	}

	/**
	 * "PrerequisiteBuild" states the specific build that
	 * the OTA package is intended for.
	 *
	 * @return The "PrerequisiteBuild" key, as a String.
     **/
	public String prerequisiteBuild() {
		return (ENTRY.containsKey("PrerequisiteBuild")) ? ENTRY.get("PrerequisiteBuild").toString() : "N/A";
	}

	/**
	 * "PrerequisiteVersion" states the specific version that
	 * the OTA package is intended for.
	 *
	 * @return The "PrerequisiteVersion" key, as a String.
     **/
	public String prerequisiteVer() {
		// Using a switch/case since exceptions must be made.
		// (I do not intend to address all exceptions.)
		switch (this.prerequisiteBuild()) {
			// iOS exceptions
			case "10A405":
				return "6.0";

			case "10B141":
				return "6.1";

			case "13A340":
			case "13A341":
				return "9.0 GM";

			// watchOS exceptions
			case "13V5098e":
				return "2.2 beta";

			case "13V5108c":
				return "2.2 beta 2";

			case "13V5117c":
				return "2.2 beta 3";

			case "13V5129c":
				return "2.2 beta 4";

			case "13V5141a":
				return "2.2 beta 5";

			case "13V5143a":
				return "2.2 beta 6";

			case "13V413":
				return "2.2.1 beta";

			case "13V601":
				return "2.2.2 beta";
		
			case "14S5247t":
				return "3.0 beta";

			case "14S5278d":
				return "3.0 beta 2";

			case "14S5290d":
				return "3.0 beta 3";

			case "14S5302d":
				return "3.0 beta 4";

			case "14S5315a":
				return "3.0 beta 5";

			case "14S5321a":
				return "3.0 beta 6";

			case "14S443":
				return "3.1 beta Pre-release";
				
			case "14S452":
				return "3.1 beta";

			case "14S464":
				return "3.1 beta 2";

			case "14S466":
				return "3.1 beta 3";

			case "14S468":
				return "3.1 beta 4 Pre-release";

			case "14S471":
				return "3.1 beta 4";

			case "14S5862d":
				return "3.1.1 beta";

			case "14S5869b":
				return "3.1.1 beta 2";

			case "14S5875b":
				return "3.1.1 beta 3";

			case "14S6879":
				return "3.1.1 beta 4";

			default:
				return (ENTRY.containsKey("PrerequisiteOSVersion")) ? ENTRY.get("PrerequisiteOSVersion").toString() : "N/A";
		}
	}

	/**
	 * Checks if Apple marked the OTA package with a release type.
	 * This function only checks for the existence of such a key,
	 * and not its value. For its value, use the betaType() method.
	 * 
	 * @return A boolean value of whether Apple provided a release type (true) or not (false).
	 **/
	public String releaseType() {
		return (ENTRY.containsKey("ReleaseType")) ? ENTRY.get("ReleaseType").toString() : "Public";
	}

	/**
	 * @return The package's file size, as a String. It is formatted with commas.
     **/
	public String size() {
		return NumberFormat.getNumberInstance(Locale.US).format(SIZE);
	}

	/**
	 * "sortingBuild()" is the regular build number, with additional zeroes
	 * in the beginning to make sure it's arranged on top.
	 *
	 * @return A String with the same value as OTAPackage.build(),
	 * but with a number of zeroes in front so the program arranges it above
	 * newer entries.
     **/
	public String sortingBuild() {
		int letterPos;
		String sortBuild = this.declaredBuild();

		// Make 9A### appear before 10A###.
		if (Character.isLetter(sortBuild.charAt(1)))
			sortBuild = '0' + sortBuild;

		// If the build number is false, replace everything after the letter with "0000."
		// This will cause betas to appear first.
		if (this.isHonestBuild() == false) {
			for (letterPos = 1; letterPos < sortBuild.length(); letterPos++) {
				if (Character.isUpperCase(sortBuild.charAt(letterPos))) {
					letterPos++;
					break;
				}
			}

			sortBuild = sortBuild.substring(0, letterPos) + "0000";
		}

		// Apple Watch betas go on the bottom in wiki markup.
		// As dumb as this is, it's a pain because of the OS version.
		// Hopefully this gets changed for consistency in the future...
		else if (this.declaredBuild().matches(REGEX_BETA)) {
			for (NSObject supportedDevice:this.supportedDevices()) {
				if (supportedDevice.toString().contains("Watch")) {
					sortBuild = sortBuild.substring(0, 3) + '9' + sortBuild.substring(4);
					break;
				}
			}
		}

		return sortBuild;
	}

	/**
	 * "sortingPrerequisiteBuild()" is the prerequisite build,
	 * with additional zeroes in the beginning to make sure
	 * it's arranged on top.
	 * 
	 * This also appends with an integer that represents the release type.
	 *
	 * @return A String with the same value as OTAPackage.prerequisiteBuild(),
	 * but with additional zeroes so the program will order it correctly,
	 * and an integer at the end specifying the release type.
     **/
	public String sortingPrerequisiteBuild() {
		// Sort by release type.
		int relType = 0;
		String build = this.prerequisiteBuild();

		switch (this.releaseType()) {
			case "Beta":
				relType = 1;
				break;

			case "Carrier":
				relType = 2;
				break;

			case "Internal":
				relType = 3;
				break;
		}

		if (this.isUniversal())
			return "000000000" + relType;

		else {
			if (Character.isLetter(build.charAt(1)))
				build = '0' + build;
			
			match = Pattern.compile("\\d?\\d[A-Z]").matcher(build);
			
			if (build.split("[A-z]")[1].length() < 3 && match.find())
				build = match.group() + '0' + build.replaceFirst("\\d?\\d[A-Z]", "");
			
			return build;
		}
	}

	/**
	 * This method provides the models that this OTA update supports.
	 *
	 * @return An array of NSObjects (which should be NSStrings).
     **/
	public NSObject[] supportedDeviceModels() {
		// Retrieve the list of supported models... if it exists.
		return (ENTRY.containsKey("SupportedDeviceModels")) ? ((NSArray)ENTRY.objectForKey("SupportedDeviceModels")).getArray() : null;
	}

	public NSObject[] supportedDevices() {
		return ((NSArray)ENTRY.objectForKey("SupportedDevices")).getArray();
	}

	/**
	 * @return The package's URL, as a String.
     **/
	public String url() {
		return URL;
	}
}

File: Octothorpe.Core/OTAPackage.cs
/*
 * Copyright (c) 2017 Dialexio
 * 
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace Octothorpe
{
    class OTAPackage
    {
        private Match match;
        private readonly Dictionary<string, object> ENTRY;

        public OTAPackage(Dictionary<string, object> package)
        {
            ENTRY = package;
        }

        /// <summary>
        /// Returns the package's actual build number. (i.e. Without any of Apple's padding.)
        /// </summary>
        /// <returns>
        /// A string value of the package's actual build number. (e.g. 10A550 will be 10A550, 12F5061 will be 12F61)
        /// </returns>
        public string ActualBuild
        {
            get
            {
                // If it the build number looks like a beta...
                // And it's labeled as a beta...
                // But it's not a beta... We need the actual build number.
                if (Regex.Match(this.DeclaredBuild, REGEX_BETA).Success &&
                    this.ReleaseType != "Public" &&
                    this.ActualReleaseType == 0)
                {
                    int letterPos, numPos;

                    for (letterPos = 1; letterPos < this.DeclaredBuild.Length; letterPos++)
                    {
                        if (char.IsUpper(this.DeclaredBuild[letterPos]))
                        {
                            letterPos++;
                            break;
                        }
                    }

                    numPos = letterPos + 1;
                    if (this.DeclaredBuild[numPos] == '0')
                        numPos++;

                    return this.DeclaredBuild.Substring(0, letterPos) + this.DeclaredBuild.Substring(numPos);
                }

                else
                    return this.DeclaredBuild;
            }
        }

        /// <summary>
        /// Returns the package's actual release type in the form of an integer.
        /// </summary>
        /// <returns>
        /// An integer value of 0 (not a beta), 1 (public beta), 2 (developer beta), 3 (carrier beta), or 4 (internal build). -1 may be returned if the type is unknown.
        /// </returns>
        public int ActualReleaseType
        {
            get
            {
                // Just check ReleaseType and return values based on it.
                switch (this.ReleaseType)
                {
                    // We do need to dig deeper if it's "Beta" though.
                    case "Beta":
                        if (this.DocumentationID == "N/A" || this.DocumentationID == "PreRelease")
                            return 2;

                        else if (this.DocumentationID.Contains("Public"))
                            return 1;

                        else if (this.DocumentationID.Contains("Beta") || this.DocumentationID.Contains("Seed"))
                            return 2;

                        else
                            return 0;

                    case "Carrier":
                        return 3;

                    case "Internal":
                        return 4;

                    case "Public":
                        return 0;

                    default:
                        Console.WriteLine("Unknown ReleaseType: " + ENTRY["ReleaseType"]);
                        return -1;
                }
            }
        }

        /// <summary>
        /// Returns the beta number of this entry. If a beta number cannot be determined, returns 0.
        /// </summary>
        /// <returns>
        /// Whatever number beta this is, as an int.
        /// </returns>
        public int BetaNumber
        {
            get
            {
                char digit = this.DocumentationID[this.DocumentationID.Length - 1];

                if (this.IsHonestBuild && Regex.IsMatch(this.DocumentationID, "(Public|Beta|Seed)"))
                    return (char.IsDigit(digit)) ? (int)char.GetNumericValue(digit) : 1;

                else
                    return 0;
            }
        }

        /// <summary>
        /// Returns the value in the entry's CompatibilityVersion key. If the compatibility version cannot be determined, this returns 0.
        /// </summary>
        /// <returns>
        /// Whatever is in the key CompatibilityVersion, as an int.
        /// </returns>
        public int CompatibilityVersion
        {
            get
            {
                return (ENTRY.ContainsKey("CompatibilityVersion")) ?
                    (int)ENTRY["CompatibilityVersion"] :
                    0;
            }
        }

        /// <summary>
        /// Returns the timestamp found in the url. Note that this is not always accurate; it may be off by a day, or even a week.
        /// </summary>
        /// <returns>
        /// The timestamp found in the url, which may not be accurate. Its format is YYYYMMDD.
        /// </returns>
        public string Date()
        {
            match = Regex.Match(this.URL, @"\d{4}(\-|\.)20\d{8}\-");

            if (match.Success)
                return match.ToString().Substring(5, 9) + match.ToString().Substring(10, 12) + match.ToString().Substring(13, 15);

            else
            {
                match = Regex.Match(this.URL, @"\d{4}(\-|\.)20\d{4}(\d|\.)(\w|\-)");

                if (match.Success)
                {
                    switch (match.ToString().Substring(5))
                    {
                        case "201218.D22":
                            return "20120307";

                        case "2015106-DC":
                            return "20151006";

                        case "20160009/1":
                            return "20160913";

                        default:
                            return match.ToString().Substring(5);
                    }
                }

                else
                    return "00000000";
            }
        }

        /// <summary>
        /// Returns part of the timestamp found in the url.
        /// </summary>
        /// <returns>
        /// The day, month, or year found in the url.
        /// </returns>
        /// <param name="dmy">Indicates if you just want the day, month, or year.</param>
        public string Date(char dmy)
        {
            switch (dmy)
            {
                // Day
                case 'd':
                    return this.Date().Substring(6);

                // Month
                case 'm':
                    return this.Date().Substring(4, 2);

                // Year
                case 'y':
                    return this.Date().Substring(0, 4);

                default:
                    return this.Date();
            }
        }

        public string DeclaredBuild
        {
            get { return (string)ENTRY["Build"]; }
        }

        /// <summary>
        /// Reports the documentation ID that Apple assigned. This tells iOS which documentation file to load, since multiple ones may be offered.
        /// </summary>
        /// <returns>
        /// The documentation ID that corresponds to the OTA update. If one is not specified, returns "N/A."
        /// </returns>
        public string DocumentationID
        {
            get
            {
                return ENTRY.ContainsKey("SUDocumentationID") ?
                    (string)ENTRY["SUDocumentationID"] :
                    "N/A";
            }
        }

        /// <summary>
        /// Checks if the release has an inflated build number. Apple does this to push devices on beta builds to stable builds.
        /// </summary>
        /// <returns>
        /// A boolean value of whether this release has a false build number (true) or not (false).
        /// </returns>
        public bool IsHonestBuild
        {
            get { return this.ActualBuild == this.DeclaredBuild; }
        }

        /// <summary>
        /// A simple check if the release is a large, "one Size fits all" package.
        /// </summary>
        /// <returns>
        /// A boolean value of whether this release is used to cover all scenarios (true) or not (false).
        /// </returns>
        public bool IsUniversal
        {
            get { return this.PrerequisiteBuild == "N/A"; }
        }

        /// <summary>
        /// Returns the value of "MarketingVersion" if present. "MarketingVersion" is used in some entries to display a false version number.
        /// </summary>
        /// <returns>
        /// A string value of the "MarketingVersion" key (if it exists), otherwise returns the "OSVersion" key.
        /// </returns>
        public string MarketingVersion
        {
            get
            {
                if (ENTRY.ContainsKey("MarketingVersion"))
                {
                    string mv = (string)ENTRY["MarketingVersion"];

                    return (mv.Contains(".") == false) ?
                        mv + ".0" :
                        mv;
                }

                else
                    return this.OSVersion;
            }
        }

        /// <returns>
        /// The "OSVersion" key, as a string. For iOS 10 and newer, this will also strip "9.9." from the version.
        /// </returns>
        public string OSVersion
        {
            get
            {
                Dictionary<string, string> VersionStrings = null;

                try
                {
                    using (StreamReader Json = File.OpenText(AppDomain.CurrentDomain.BaseDirectory + "OS versions.json"))
                    {
                        VersionStrings = JsonConvert.DeserializeObject<Dictionary<string, string>>(Json.ReadToEnd());
                        return VersionStrings[this.ActualBuild];
                    }
                }

                catch (KeyNotFoundException)
                {
                    string version = (string)ENTRY["OSVersion"];
                    return (version.Substring(0, 3) == "9.9") ? version.Substring(4) : version;
                }
            }
        }

        /// <summary>
        /// "PrerequisiteBuild" states the specific build that the OTA package is intended for, since most OTA packages are deltas.
        /// </summary>
        /// <returns>
        /// The "PrerequisiteBuild" key, as a string.
        /// </returns>
        public string PrerequisiteBuild
        {
            get
            {
                return (ENTRY.ContainsKey("PrerequisiteBuild")) ?
                    (string)ENTRY["PrerequisiteBuild"] :
                    "N/A";
            }
        }

        /// <summary>
        /// "PrerequisiteVersion()" states the specific version that the OTA package is intended for, since most OTA packages are deltas.
        /// </summary>
        /// <returns>
        /// The "PrerequisiteVersion" key, as a string.
        /// </returns>
        
        public string PrerequisiteVer
        {
            get
            {
                Dictionary<string, string> VersionStrings = null;
                
                try
                {
                    using (StreamReader Json = File.OpenText(AppDomain.CurrentDomain.BaseDirectory + "OS versions.json"))
                    {
                        VersionStrings = JsonConvert.DeserializeObject<Dictionary<string, string>>(Json.ReadToEnd());
                        return VersionStrings[this.PrerequisiteBuild];
                    }
                }
                
                catch (KeyNotFoundException)
                {
                    return (ENTRY.ContainsKey("PrerequisiteOSVersion")) ?
                        (string)ENTRY["PrerequisiteOSVersion"] :
                        "N/A";
                }
            }
        }

        /// <summary>
        /// This provides an easily accessible regular expression to detect build numbers that are (likely to be) beta versions.
        /// </summary>
        /// <returns>
        /// A regular expression, as a string, that can be used to detect build numbers belonging to beta versions.
        /// </returns>
        public static string REGEX_BETA
        {
            get { return @"(\d)?\d[A-Z][4-6]\d{3}[a-z]?"; }
        }

        /// <summary>
        /// Checks if Apple marked the OTA package with a release type. This function returns its value, but it may not be accurate. If accuracy is needed, use the ActualReleaseType() method.
        /// </summary>
        /// <returns>
        /// The "ReleaseType" key, as a string. If the key is not present, returns "Public."
        /// </returns>
        public string ReleaseType
        {
            get
            {
                return (ENTRY.ContainsKey("ReleaseType")) ?
                    (string)ENTRY["ReleaseType"] :
                    "Public";
            }
        }

        /// <summary>
        /// Removes padding from a build number, supplied as a string argument.
        /// </summary>
        /// <returns>
        /// The build number, minus the padding for a beta.
        /// </returns>
        private string RemoveBetaPadding(string BuildNum)
        {
            if (Regex.Match(this.DeclaredBuild, REGEX_BETA).Success)
            {
                int LetterPos;

                for (LetterPos = 1; LetterPos < BuildNum.Length; LetterPos++)
                {
                    if (char.IsUpper(BuildNum[LetterPos]))
                    {
                        LetterPos++;
                        break;
                    }
                }

                return BuildNum.Substring(0, LetterPos) + BuildNum.Substring(LetterPos + 1);
            }

            else
                return BuildNum;
        }

        /// <summary>
        /// This is the size of the ZIP file.
        /// </summary>
        /// <returns>
        /// Returns the size of the OTA package as a number, formatted with commas.
        /// </returns>
        public string Size
        {
            get
            {
                if (ENTRY.ContainsKey("RealUpdateAttributes"))
                {
                    var RealUpdateAttrs = (Dictionary<string, object>)ENTRY["RealUpdateAttributes"];
                    return string.Format("{0:n0}", RealUpdateAttrs["RealUpdateDownloadSize"]);
                }

                else
                    return string.Format("{0:n0}", ENTRY["_DownloadSize"]);
            }
        }

        /// <summary>
        /// This string is used for sorting purposes.
        /// </summary>
        /// <returns>
        /// Returns the values of ActualReleaseType(), a padded build, and a padded prerequisite build number in that order.
        /// </returns>
        public string SortingString
        {
            get { return this.SortingBuild() + this.SortingPrerequisiteBuild() + this.CompatibilityVersion + this.ActualReleaseType; }
        }

        private string SortingBuild()
        {
            int LetterPos;
            string SortBuild = this.DeclaredBuild;

            // Make 9A### appear before 10A###.
            if (char.IsLetter(SortBuild[1]))
                SortBuild = '0' + SortBuild;

            // If the build number is false, replace everything after the letter with "0000."
            // This will cause betas to appear first.
            if (this.IsHonestBuild == false)
            {
                for (LetterPos = 1; LetterPos < SortBuild.Length; LetterPos++)
                {
                    if (char.IsUpper(SortBuild[LetterPos]))
                    {
                        LetterPos++;
                        break;
                    }
                }

                SortBuild = SortBuild.Substring(0, LetterPos) + "0000";
            }

            else if (Regex.Match(this.DeclaredBuild, REGEX_BETA).Success)
                SortBuild = RemoveBetaPadding(SortBuild);

            return SortBuild;
        }

        private string SortingPrerequisiteBuild()
        {
            // Sort by release type.
            int ReleaseTypeInt = 0;
            string build = this.PrerequisiteBuild;

            // Get up to (and including) the first letter. We'll get to this in a bit.
            match = Regex.Match(build, @"\d?\d[A-Z]");

            switch (this.ReleaseType)
            {
                case "Beta":
                    ReleaseTypeInt = 1;
                    break;

                case "Carrier":
                    ReleaseTypeInt = 2;
                    break;

                case "Internal":
                    ReleaseTypeInt = 3;
                    break;
            }

            if (this.IsUniversal)
                return "000000000" + ReleaseTypeInt;

            // If the build is old (i.e. before iOS 7), pad it.
            if (char.IsLetter(build[1]))
                return '0' + build;
                
            if (this.PrerequisiteVer.Contains("beta"))
                build = RemoveBetaPadding(build);

            // If the number after the capital letter is too small, pad it.
            if (new Regex("[A-z]").Split(build)[1].Length < 3 && match.Success)
                build = match.Value + '0' + new Regex(@"\d?\d[A-Z]").Replace(build, "", 1);

            // If the build does not have a letter, add a fake one to push it below similarly-numbered betas.
            if (char.IsLetter(build[build.Length-1]))
                build = build + 'z';

            return build;
        }

        /// <summary>
        /// Contains a list of all supported models.
        /// </summary>
        /// <returns>
        /// A List of strings specifying what models are supported.
        /// </returns>
        public List<string> SupportedDeviceModels
        {
            get
            {
                List<string> Models = new List<string>();

                try
                {
                    foreach (object Model in (object[])ENTRY["SupportedDeviceModels"])
                        Models.Add((string)Model);
                }

                // No models specified. (Older PLISTs do this.)
                catch (KeyNotFoundException)
                { }

                return Models;
            }
        }

        /// <summary>
        /// Contains a list of all supported devices.
        /// </summary>
        /// <returns>
        /// A List of objects (which should be strings specifying what devices are supported).
        /// </returns>
        public List<string> SupportedDevices
        {
            get
            {
                List<string> Devices = new List<string>();

                foreach (object Device in (object[])ENTRY["SupportedDevices"])
                    Devices.Add((string)Device);

                return Devices;
            }
        }

        /// <returns>
        /// The package URL, as a string.
        /// </returns>
        public string URL
        {
            get
            {
                if (ENTRY.ContainsKey("RealUpdateAttributes"))
                {
                    var RealUpdateAttrs = (Dictionary<string, object>)ENTRY["RealUpdateAttributes"];
                    return (string)RealUpdateAttrs["RealUpdateURL"];
                }

                else
                    return (string)ENTRY["__BaseURL"] + (string)ENTRY["__RelativePath"];
            }
        }
    }
}
